pkg/controller/direct/workstations/workstationcluster_controller.go

```diff
@@ -0,0 +1,315 @@
+	workstationspb "cloud.google.com/go/workstations/apiv1/workstationspb"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "workstations-controller"
+	serviceDomain = "//workstations.googleapis.com"
+)
+
+func init() {
+	registry.RegisterModel(krm.WorkstationClusterGVK, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &model{config: *config}, nil
+}
+
+var _ directbase.Model = &model{}
+
+type model struct {
+	config config.ControllerConfig
+}
+
+func (m *model) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building WorkstationCluster client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.WorkstationCluster{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	projectRef, err := refs.ResolveProject(ctx, reader, obj, &obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+
+	// Get location
+	location := obj.Spec.Location
+
+	var id *WorkstationClusterIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(projectID, location, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.Parent.Project != projectID {
+			return nil, fmt.Errorf("WorkstationCluster %s/%s has spec.projectRef changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Project, projectID)
+		}
+		if id.Parent.Location != location {
+			return nil, fmt.Errorf("WorkstationCluster %s/%s has spec.location changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Location, location)
+		}
+		if id.WorkstationCluster != resourceID {
+			return nil, fmt.Errorf("WorkstationCluster  %s/%s has metadata.name or spec.resourceID changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.WorkstationCluster, resourceID)
+		}
+	}
+
+	if err := NormalizeWorkstationCluster(ctx, reader, obj); err != nil {
+		return nil, err
+	}
```

Comment:
Suggest moving this to create/update if the refs are optional?

Reply:
The refs are required.